### PR TITLE
Adds a new `ResPath` struct that will replace `ResourcePath`

### PR DESCRIPTION
**Pull request by Ygg01**
Wanted some feedback from maintainers. 

Stuff to do:

- [x] Benchmarks
- [x] Decide on `ResPathSpan` a wrapper struct for `ReadOnlySpan`.
- [x] Decide what to do with construction. Keep same constructor as `ResourcePath`. Decided: No.
- [x] ~~Fast creation - via factory method.~~ Pointless from talki with PJB
- [x] ~~Avoid branching in fast factory method?~~ Yes.
- [x] Docs.
- [x] Tests. 

This relates to #2540
It doesn't fix because it would require wholesale replacement. 

No longer a breaking change.

### CODE REVIEW

**Comment by mirrorcult** [comment]
if this PR isnt gonna be breaking should at least obsolete `ResourcePath`

**Review by mirrorcult — APPROVED** [review]
looks fine to me especially with the test but need benchmarks imo

**Comment by Ygg01** [comment]
Added Benchmarks and fix some issues using a custom enumerator.

```
|           Method |   N |        Mean |     Error |    StdDev |
|----------------- |---- |------------:|----------:|----------:|
|      ReadResPath |   1 |    349.3 ns |   6.78 ns |   9.05 ns |
|  ReadResPathFast |   1 |    186.7 ns |   3.75 ns |   5.14 ns |
| ReadResourcePath |   1 |    470.0 ns |   9.37 ns |  15.92 ns |
|      ReadResPath |  10 |  1,864.6 ns |  36.71 ns |  64.29 ns |
|  ReadResPathFast |  10 |    265.9 ns |   5.23 ns |   8.14 ns |
| ReadResourcePath |  10 |  2,572.3 ns |  50.37 ns |  51.73 ns |
|      ReadResPath | 100 | 17,050.7 ns | 332.89 ns | 444.40 ns |
|  ReadResPathFast | 100 |  1,075.6 ns |  21.23 ns |  34.87 ns |
| ReadResourcePath | 100 | 23,655.3 ns | 351.22 ns | 328.53 ns |
```

File: Robust.Shared/Utility/ResourcePath.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@ namespace Robust.Shared.Utility
     ///     ResourcePaths are immutable.
     /// </summary>
     [PublicAPI, Serializable, NetSerializable]
+    [Obsolete]
     public sealed class ResourcePath : IEquatable<ResourcePath>
```
Comment (AJCM-git) [inline Robust.Shared/Utility/ResourcePath.cs]: Obsolete the public fields/methods in the class with a message mentioning what their replacement in ResPath is

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Robust.Shared/Utility/ResourcePath.cs (hunk 1 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResourcePath.cs]: It's the same API, albeit constructors are different.

**Comment by Ygg01** [comment]
Removed branching from the fast constructor. Added type serializer.

Speed is now: 
|                     Method |    N |           Mean |         Error |        StdDev |
|--------------------------- |----- |---------------:|--------------:|--------------:|
| CreateWithSeparatorResPath |   10 |   1,061.489 ns |     9.4685 ns |     8.8568 ns |
|    CreateUnsafePathResPath |   10 |       1.941 ns |     0.0243 ns |     0.0215 ns |
|         CreateResourcePath |   10 |   1,212.210 ns |    15.5250 ns |    12.9641 ns |
| CreateWithSeparatorResPath |  100 |  11,183.635 ns |   213.7302 ns |   178.4744 ns |
|    CreateUnsafePathResPath |  100 |      25.374 ns |     0.3536 ns |     0.3307 ns |
|         CreateResourcePath |  100 |  12,298.120 ns |   240.9642 ns |   353.2023 ns |
| CreateWithSeparatorResPath | 1000 | 106,719.099 ns | 1,165.9764 ns | 1,090.6550 ns |
|    CreateUnsafePathResPath | 1000 |     218.513 ns |     3.4429 ns |     3.2205 ns |
|         CreateResourcePath | 1000 | 127,124.898 ns | 2,526.0949 ns | 3,457.7472 ns |


However, it will allocate a bunch `CreateWithSeparator(_path)` allocates around 351,426.1 MB
while `new ResourcePath(_path)` allocates around 635,312.3 MB.

Some pooling or ref struct could help with this. 

This is why I made factory methods rather than constructors, factory methods could manage an internal pool or a `ResourceCache`.

File: Robust.Benchmarks/Utility/ResourcePathBench.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+namespace Robust.Benchmarks.Utility;
+
+[Virtual]
+public class ResourcePathBench
+{
+    private string _path = default!;
+
+    [UsedImplicitly]
+    [Params(10, 100, 1000)]
+    public int N;
```
Comment (AlexMorgan3817) [inline Robust.Benchmarks/Utility/ResourcePathBench.cs:16]: Bad naming and no `///<summary>`, try to rename to something that at least hints on explanation of what is it

File: Robust.Shared/Utility/ResPath.cs [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+    ///     <code>
+    ///     var path1 = new ResourcePath("/a/b/c");
+    ///     var path2 = new ResourcePath("/a");
+    ///     Console.WriteLine(path1.RelativeTo(path2)); // prints "b/c".
+    ///     </code>
+    /// </example>
+    /// <exception cref="ArgumentException">Thrown if we are not relative to the base path.</exception>
+    public ResPath RelativeTo(ResPath basePath)
+    {
+        if (TryRelativeTo(basePath, out var relative)) return relative.Value;
```
Comment (AlexMorgan3817) [inline Robust.Shared/Utility/ResPath.cs]:
```suggestion
        if (TryRelativeTo(basePath, out var relative))
                return relative.Value;
```

File: Robust.Shared/Utility/ResPath.cs [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+
+    /// <summary>
+    ///     Turns the path into a relative path by removing the root separator, if any.
+    ///     Does nothing if the path is already relative.
+    /// </summary>
+    /// <seealso cref="IsRelative" />
+    /// <seealso cref="ToRootedPath" />
+    public ResPath ToRelativePath()
+    {
+        if (IsRelative) return this;
```
Comment (AlexMorgan3817) [inline Robust.Shared/Utility/ResPath.cs]:
```suggestion
        if (IsRelative)
                return this;
```

File: Robust.Shared/Utility/ResPath.cs [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+        return true;
+    }
+
+    /// <summary>
+    ///     Turns the path into a relative path with system-specific separator.
+    ///     For usage in disk I/O.
+    /// </summary>
+    public string ChangeSeparator(string newSeparator)
+    {
+        if (newSeparator is "." or "\0") throw new ArgumentException("New separator can't be `.` or `NULL`");
```
Comment (AlexMorgan3817) [inline Robust.Shared/Utility/ResPath.cs]:
```suggestion
        if (newSeparator is "." or "\0")
                throw new ArgumentException("New separator can't be `.` or `NULL`");
```
Too long line, better to linebreak

**Review by AlexMorgan3817 — CHANGES_REQUESTED** [review]

**Comment by AlexMorgan3817** [comment]
More of formatting. And... Pendulum is based

File: Robust.Shared/Utility/ResPath.cs [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+    /// <summary>
+    ///     The separator for the file system of the system we are compiling to.
+    ///     Backslash on Windows, forward slash on sane systems.
+    /// </summary>
+#if WINDOWS
+    public const char SystemSeparator = '\\';
+
+    public const string SystemSeparatorStr = @"\";
+#else
+    public const char SystemSeparator = '/';
```
Comment (AlexMorgan3817) [inline Robust.Shared/Utility/ResPath.cs:25]:
```suggestion
    public const char SystemSeparator = '/';

```

File: Robust.Benchmarks/Utility/ResourcePathBench.cs (hunk 2 above)
Comment (DrSmugleaf) [reply Robust.Benchmarks/Utility/ResourcePathBench.cs:16]: N is commonly used for the amount of iterations in a benchmark
https://github.com/dotnet/BenchmarkDotNet

Comment (AlexMorgan3817) [reply Robust.Benchmarks/Utility/ResourcePathBench.cs:16]: > N is commonly used for the amount of iterations in a benchmark https://github.com/dotnet/BenchmarkDotNet

Yes, but what is iterations, do you just iterate integer numbers? :\

Comment (AlexMorgan3817) [reply Robust.Benchmarks/Utility/ResourcePathBench.cs:16]: Okay, anyway, maybe i am just not experienced in C# enough

File: Robust.Shared/Utility/ResPath.cs (hunk 6 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResPath.cs:25]: I don't see any changes here. I added a new line after block.

File: Robust.Shared/Utility/ResPath.cs [hunk 7]
```diff
@@ -22,12 +22,12 @@ namespace Robust.Shared.Utility;
     /// </summary>
 #if WINDOWS
     public const char SystemSeparator = '\\';
-
     public const string SystemSeparatorStr = @"\";
 #else
     public const char SystemSeparator = '/';
     public const string SystemSeparatorStr = "/";
 #endif
```
Comment (AlexMorgan3817) [inline Robust.Shared/Utility/ResPath.cs]:
```suggestion
    public const char SystemSeparator = '/';

    public const string SystemSeparatorStr = "/";
#endif
```
As you wish

File: Robust.Shared/Utility/ResPath.cs [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+            if (needsSeparator) sb.Append('/');
+
+            sb.Append(segment);
+            needsSeparator = true;
+        }
+
+        return sb.Length == 0
+            ? Self
+            : new ResPath(sb.ToString());
+    }
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResPath.cs]: Is CreateWIthSeparator an accurate name anymore?

File: Robust.Shared/Utility/ResPath.cs [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+    ///     No special indication is given for rooted paths,
+    ///     so <c>/a/b</c> yields the same as <c>a/b</c>.
+    /// </remarks>
+    public IEnumerable<string> EnumerateSegments()
+    {
+        foreach (var fragment in CanonPath.Split(Separator))
+        {
+            yield return fragment;
+        }
+    }
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResPath.cs]: Should be a struct enumerator.

File: Robust.Shared/Utility/ResPath.cs [hunk 10]
```diff
@@ -0,0 +1,680 @@
...
+            : new ResPath(sb.ToString());
+    }
+
+    /// <summary>
+    ///     Check whether a path is clean, i.e. <see cref="Clean"/> would not modify it.
+    /// </summary>
+    /// <returns>if true if path wouldn't be modified </returns>
+    public bool IsClean()
+    {
+        var segments = CanonPath.Split(Separator).ToArray();
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResPath.cs]: Is the ToArray needed?

File: Robust.Shared/Utility/ResPath.cs [hunk 11]
```diff
@@ -0,0 +1,680 @@
...
+        for (var i = 0; i < segments.Length; i++)
+        {
+            if (segments[i] != "..") continue;
+
+            if (IsRooted)
+            {
+                return false;
+            }
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResPath.cs]: Is this supposed to be in the loop?

File: Robust.Shared/Utility/ResPath.cs [hunk 12]
```diff
@@ -0,0 +1,680 @@
...
+
+    /// <summary>
+    ///     Returns cleaned version of the resource path, removing <c>..</c>.
+    /// </summary>
+    /// <remarks>
+    ///     If <c>..</c> appears at the base of a path, it is left alone. If it appears at root level (like <c>/..</c>) it is removed entirely.
+    /// </remarks>
+    public ResPath Clean()
+    {
+        var segments = new List<string>();
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResPath.cs]: ValueList, should also be constructed after the if check.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
PJB should also take a look.

File: Robust.Shared/Utility/ResPath.cs (hunk 8 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResPath.cs]: It's a placeholder name for constructor taking path with arbitrary separators. E.g. `a|b|c` and separator `|` result in `a/b/c` `CanonPath`.

File: Robust.Shared/Utility/ResPath.cs (hunk 11 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResPath.cs]: Honestly, it's verbatim of `ResourcePath` method, I really had no luck figuring it out at the time. That said, I maybe could try and refactor it, possibly remove the array.

File: Robust.Shared/Utility/ResPath.cs (hunk 9 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResPath.cs]: Meet you halfway and make it return `string[]` it won't break interfaces.

**Review by PJB3005 — CHANGES_REQUESTED** [review]
Much of the complexity kept from `ResourcePath` isn't necessary and needs to be removed. The maintainer meeting for this was a year ago but we agreed all the complexities with "regular filesystem" semantics like `.` and `..` traversal should just be axed. Really the only thing you need at most is like, concatenation taking root vs relative into account.

File: Robust.Shared/Utility/ResPath.cs (hunk 8 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResPath.cs]: Removed per PJB comment.

File: Robust.Shared/Utility/ResPath.cs (hunk 9 above)
Comment (Ygg01) [reply Robust.Shared/Utility/ResPath.cs]: Removed per PJB's comment.

File: Robust.Shared/Serialization/TypeSerializers/Implementations/ResPathSerializer.cs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            path /= "meta.json";
+        }
+
+        if (!path.CanonPath.Split('/').First().Equals("Textures", StringComparison.InvariantCultureIgnoreCase))
+        {
+            path = SharedSpriteComponent.TextureRootResPath / path;
+        }
+
+        path = path.ToRootedPath();
```
Comment (metalgearsloth) [inline Robust.Shared/Serialization/TypeSerializers/Implementations/ResPathSerializer.cs:34]: Is keeping /Textures support worth it.

File: Robust.Shared/Utility/ResPath.cs [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+
+    /// <summary>
+    ///     "/" (root) as a static. Separator used is <c>/</c>.
+    /// </summary>
+    public static readonly ResPath Root = new("/");
+
+    /// <summary>
+    ///     "/" (root) as a static. Separator used is <c>/</c>.
+    /// </summary>
+    public static readonly ResPath Empty = new("");
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResPath.cs]: doc summary looks copied?

File: Robust.Shared/Utility/ResourcePath.cs [hunk 15]
```diff
@@ -15,6 +15,8 @@ namespace Robust.Shared.Utility
     ///     ResourcePaths are immutable.
     /// </summary>
     [PublicAPI, Serializable, NetSerializable]
+    [Obsolete(@$"Use {nameof(ResPath)} instead, for extra performance
+        use {nameof(ResPath)} but only if '/' is the separator")]
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/ResourcePath.cs]: Comment looks doubled? Should just say to use ResPath instead.

**Review by metalgearsloth — APPROVED** [review]

File: Robust.Shared/Serialization/TypeSerializers/Implementations/ResPathSerializer.cs (hunk 13 above)
Comment (Ygg01) [reply Robust.Shared/Serialization/TypeSerializers/Implementations/ResPathSerializer.cs:34]: Shrug, no idea; I copied to make sure serialization worked in-game (when I briefly replaced `ResourcePath` with `ResPath`).

https://github.com/space-wizards/RobustToolbox/blob/40c8a3ccf78bd3f9b8f9554242e3bf373c968f89/Robust.Shared/Serialization/TypeSerializers/Implementations/ResourcePathSerializer.cs#L40-L43

**Pull request merged**